### DISCUSSION TITLE
YOLOv8-Pose CPP Inference with Libtorch

### DISCUSSION THREAD

**Discussion by masgura** [ultralytics/ultralytics#8754] [Q&A]
Hi,

I'm trying to do inference with YOLOv8-Pose on C++, using the Pytorch C++ API. I'm referring to the example found in https://github.com/ultralytics/ultralytics/blob/main/examples/YOLOv8-LibTorch-CPP-Inference/main.cc, but I'm having an hard time trying to modify the output of the non max suppression function to include also the keypoints coordinate. Does anybody have an idea on how to do it?

**Comment by pderrenger** [comment]
Hi there! 👋

Great to hear you're exploring YOLOv8-Pose with C++ and Libtorch! To include keypoints in the output of the non-max suppression (NMS) function, you'll need to adjust the output tensor structure to accommodate the keypoints data.

Assuming your model outputs keypoints in a separate tensor, you can modify the NMS function to return an additional tensor for keypoints. Here's a simplified example of how you might adjust the function:

```cpp
auto outputs = model.forward(inputs).toTuple();
auto detections = outputs->elements()[0].toTensor();
auto keypoints = outputs->elements()[1].toTensor();

// Apply NMS to detections
auto nms_results = non_max_suppression(detections, conf_thres, iou_thres);

// Now, you need to correlate nms_results with keypoints
// This part depends on how your model outputs keypoints and detections
```

This example assumes your model's forward method returns a tuple with detections and keypoints. You'll need to adjust the non-max suppression function to handle the keypoints tensor alongside detections.

Remember, the exact implementation will depend on your model's output format and how keypoints are represented. If keypoints are included within the detection tensor, you'll need to extract them accordingly.

Hope this helps! Let us know if you have further questions. 😊
